src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1245,33 +1260,55 @@ impl Store {
     &self,
     destination: PathBuf,
     digest: DirectoryDigest,
+    mutable_paths: &BTreeSet<RelativePath>,
+    immutable_inputs: Option<&ImmutableInputs>,
     perms: Permissions,
   ) -> Result<(), StoreError> {
     // Load the DigestTrie for the digest, and convert it into a mapping between a fully qualified
     // parent path and its children.
     let mut parent_to_child = HashMap::new();
-    self
-      .load_digest_trie(digest)
-      .await?
-      .walk(SymlinkBehavior::Aware, &mut |path, entry| {
-        if let Some(parent) = path.parent() {
-          parent_to_child
-            .entry(destination.join(parent))
-            .or_insert_with(Vec::new)
-            .push(entry.clone());
-        }
-      });
+    let tree = self.load_digest_trie(digest).await?;
+    tree.walk(SymlinkBehavior::Aware, &mut |path, entry| {
+      if let Some(parent) = path.parent() {
+        parent_to_child
+          .entry(destination.join(parent))
+          .or_insert_with(Vec::new)
+          .push(entry.clone());
+      }
+    });
+
+    let mut mutable_path_ancestors = BTreeSet::new();
+    for relpath in mutable_paths {
+      mutable_path_ancestors.extend(relpath.ancestors().map(|p| destination.join(p)));
+    }
+    mutable_path_ancestors.remove(&destination);
 
     self
-      .materialize_directory_helper(destination, true, &parent_to_child, perms)
+      .materialize_directory_children(
+        destination.clone(),
+        true,
+        false,
+        &parent_to_child,
+        &mutable_paths
+          .iter()
+          .map(|p| destination.join(p))
+          .collect::<BTreeSet<_>>(),
+        &mutable_path_ancestors,
+        immutable_inputs,
+        perms,
+      )
       .await
   }
 
-  fn materialize_directory_helper<'a>(
+  fn materialize_directory_children<'a>(
     &self,
     destination: PathBuf,
     is_root: bool,
+    force_mutable: bool,
     parent_to_child: &'a HashMap<PathBuf, Vec<directory::Entry>>,
+    mutable_paths: &'a BTreeSet<PathBuf>,
+    mutable_path_ancestors: &'a BTreeSet<PathBuf>,
```

Comment:
I'm not certain that you need both of these sets. I think that the actual behavior is the same for both ancestors and actual mutable paths: the only thing must be treated as mutable. So can just have one set?

Reply:
Ah we needed both when we had a directory symlinking, but now that this is only file, we don't... I think...